### PR TITLE
Split terminal widget(browser) logic

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### Referenced issue:
https://github.com/theia-ide/theia/issues/4523
https://github.com/eclipse/che/issues/12609

Split terminal widget(browser) logic

### Requested changes:
Split terminal widget logic into two components: TerminalWidget and TerminalClient:
 - TerminalWidget - contains logic to configure and create new terminal widget and methods to manipulate with this widget.
 - TerminalClient - create connection with terminal backend and manage terminal widget; encapsulate theia backend connection specific logic(to have ability simplier rebind terminal widget to non default back-end).
3. Remove using in the extensions depracated method activateTerminal.
4. Resolve an issue : task widget should display title with label task, not terminal!!!
5. Simplify imports for @theia/terminal and @theia/xterm.js browser package.

### Breaking stuff:
Depracate terminalWidget.start() method. This method has too tricky behavior. I think this methods does too much operations with too much cases. I propose to split this method onto two separated: 
1) createAndAttach()
2) attach()

### Benefits(or why do we need this...) 
1. Split code base.
2. It's useful for terminal plugin api. We could provide in the **plugin api** method to create new **terminal widget without connection to the Theia specific backend**. And user could define own connection logic(to the own backend) in the Theia plugin.

Notice for future: this pull request it's a first step: split logic. Second step(next pull request): introduce ability to create terminal widget with help plugin api.

### CODE REVIEW

**Comment by akosyakov** [comment]
Don't you xterm anymore? Why a new extension is created for it?

File: packages/terminal/src/browser/base/terminal-client.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+     * target process was not found by terminalId or it's gone. NOTICE: False by default.
+     */
+    attach(terminalId: number, createNewTerminalOnFail?: boolean): Promise<number>;
+
+    /**
+     * Resize terminal process on the backend side.
+     * @param cols - amount displayed columns in the output frame.
+     * @param rows - amount displayed rows in the output frame.
+     */
+    resize(size: TerminalSize): Promise<void>;
```
Comment (akosyakov) [inline packages/terminal/src/browser/base/terminal-client.ts]: it is never used in `TerminalWidget`, why is it exposed?

File: packages/terminal/src/browser/base/terminal-client.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Kill backend process.
+     */
+    kill(): Promise<void>;
```
Comment (akosyakov) [inline packages/terminal/src/browser/base/terminal-client.ts]: it's never used in  `TerminalWidget`. Why is it exposed?

File: packages/terminal/src/browser/default-terminal-client.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                this.logger.error(`The terminal process terminated. Cause: ${error}`);
+            }
+        }));
+        this.toDispose.push(this.terminalWatcher.onTerminalExit(({ terminalId }) => {
+            if (terminalId === this.terminalId) {
+                this.dispose();
+            }
+        }));
+
+       this.toDispose.push(this.widget.onTerminalResize(size => this.resize(size)));
```
Comment (akosyakov) [inline packages/terminal/src/browser/default-terminal-client.ts]: Should not this code always be in `TerminalWidget`?

**Comment by akosyakov** [comment]
> We could provide in the plugin api method to create new terminal widget without connection to the Theia specific backend.

Could you elaborate on it? Do you want to connect to an alternative backend in Eclipse Che?

**Comment by benoitf** [comment]
For now in Eclipse Che there is a fork of the Theia terminal extension. (To have multi remote terminal access). The idea is to move some logic to be re-used so it can work with different endpoints like  endpoint = host running theia or a different endpoint (all containers in Eclipse Che).

**Comment by AndrienkoAleksandr** [comment]
> Don't you use xterm anymore? Why a new extension is created for it? It seems this PR is attempt to abstract away connection logic, not a concrete terminal implementation.
I've never liked a separation between TerminalWidget and TerminalWidgetImpl. This PR takes separation one step farther without any reasons. It could be as simple like TerminalWidget (one class with interface and implementation parts) -> Terminal (interface) if we wanted to abstract a concrete terminal.
cc @svenefftinge

If you against this changes, I could revert it. But for me it's not bad if somebody with minimal changes could integrate some another terminal widget.

**Comment by AndrienkoAleksandr** [comment]
> It seems this PR is attempt to abstract away connection logic, not a concrete terminal implementation.

Yes, it's really "attempt" to split connection logic. And I didn't intergrate all wanted changes for better split, because for that I need change terminal restore logic, it's possible, but it could be buggy. So for this step I did minimal changes for split logic + moving xterm.js to the separated extension.

**Comment by AndrienkoAleksandr** [comment]
> Don't you use xterm anymore?

We still want to use xterm.js. It's nice widget, and it become better and better.

File: packages/terminal/src/browser/default-terminal-client.ts (hunk 3 above)
Comment (AndrienkoAleksandr) [reply packages/terminal/src/browser/default-terminal-client.ts]: Could You elaborate more information, please? You against dispose of resize subscription? Or Do you want that terminal widget call TerminalClient#resize?

**Comment by akosyakov** [comment]
> If you against this changes, I could revert it. But for me it's not bad if somebody with minimal changes could integrate some another terminal widget #1204 .

The point is that it should be done differently. `TerminalWidget` should be the same, there should be a new `Terminal` interface used by  `TerminalWidget`. xterm/domtermn extensions should only implement `Terminal` interface. It is how it is done for editor. Not like in this PR.

**Comment by AndrienkoAleksandr** [comment]
> please make this PR about it and revert all other unnecessary changes

Reverted xterm.js extension. Should I revert something else?

https://github.com/theia-ide/theia/pull/4518#discussion_r264102306 sorry, apply more information, please, I didn't understood what did you mean.

File: packages/terminal/src/browser/terminal-widget-impl.ts [hunk 4]
```diff
@@ -53,37 +49,43 @@ interface TerminalCSSProperties {
...
     @inject(ThemeService) protected readonly themeService: ThemeService;
     @inject(ILogger) @named('terminal') protected readonly logger: ILogger;
     @inject('terminal-dom-id') public readonly id: string;
     @inject(TerminalPreferences) protected readonly preferences: TerminalPreferences;
 
+    @inject(TerminalClientFactory) terminalClientFactory: TerminalClientFactory;
+
+    private terminalClient: TerminalClient;
+
+    protected readonly _onTerminalDidClose = new Emitter<TerminalWidget>();
```
Comment (akosyakov) [inline packages/terminal/src/browser/terminal-widget-impl.ts]: Remove, renaming protected properties is a breaking change, please keep names there it is possible.

File: packages/terminal/src/browser/default-terminal-client.ts (hunk 3 above)
Comment (akosyakov) [reply packages/terminal/src/browser/default-terminal-client.ts]: I meant that https://github.com/theia-ide/theia/blob/7cea837f5cf9607c86cc2f65a6799e0742f4ba3b/packages/terminal/src/browser/default-terminal-client.ts#L80-L93 could stay in `TerminalWidgetImpl`.

File: packages/terminal/src/browser/default-terminal-client.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+    get terminalId(): number {
+        return this._terminalId;
+    }
+
+    @postConstruct()
+    protected init(): void {
+        this.toDispose.push(this.terminalWatcher.onTerminalError(({ terminalId, error }) => {
+            if (terminalId === this.terminalId) {
+                this.dispose();
```
Comment (akosyakov) [inline packages/terminal/src/browser/default-terminal-client.ts]: before it will always fire close event, the same for `onTerminalExit`

**Comment by akosyakov** [comment]
> Reverted xterm.js extension. Should I revert something else?

the change should be minimal as possible with minimal breaking changes. terminal is quite central thing to break it.

File: packages/terminal/src/browser/base/terminal-client.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * Terminal client options to setup backend terminal process
+     * and control terminal client configuration.
+     */
+    readonly options: TerminalClientOptions;
+
+    /**
+     * Terminal widget which is under TerminalClient control.
+     */
+    readonly widget: TerminalWidget;
```
Comment (akosyakov) [inline packages/terminal/src/browser/base/terminal-client.ts]: that's very weird that there is a cycle dependency between `TerminalClient` and `TerminalWidget`

File: packages/terminal/src/browser/base/terminal-client.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    createAndAttach(): Promise<number>;
+
+    /**
+     * Attach to already running process.
+     * @param terminalId - unique process backend id(it's not PID, it's backend id to count runnting processes).
+     * @param createNewTerminalOnFaill - spawn new process in case if
+     * target process was not found by terminalId or it's gone. NOTICE: False by default.
+     */
+    attach(terminalId: number, createNewTerminalOnFail?: boolean): Promise<number>;
```
Comment (akosyakov) [inline packages/terminal/src/browser/base/terminal-client.ts]: i would use an object as a second arg to provide extensible options

File: packages/terminal/src/browser/base/terminal-widget.ts [hunk 8]
```diff
@@ -17,27 +17,76 @@
...
 
+    /*
+     * Send event on user text typing.
+     */
+    abstract onUserInput: Event<string | undefined>;
+
+    /*
+     * Send event on terminal resize.
+     */
+    abstract onTerminalResize: Event<TerminalSize>;
```
Comment (akosyakov) [inline packages/terminal/src/browser/base/terminal-widget.ts]: `on[Did/Will]Verb(Noun)?` is how events should be named

File: packages/terminal/src/browser/terminal-widget-impl.ts [hunk 9]
```diff
@@ -53,37 +49,43 @@ interface TerminalCSSProperties {
...
+    protected readonly _onUserInput = new Emitter<string | undefined>();
```
Comment (akosyakov) [inline packages/terminal/src/browser/terminal-widget-impl.ts]: also please study naming conventions, we use `_` only in some cases

**Comment by AndrienkoAleksandr** [comment]
Close like to much outdated. And VSCode already has such thing like we wanted https://github.com/microsoft/vscode/blob/master/src/vs/vscode.d.ts#L7518

**Pull request closed without merging**